In [2]:
import requests
import re
from bs4 import BeautifulSoup
from bs4.element import Comment
import string
import pickle

In [3]:
def get_url_in_response(text):
    urls = []
    urls = re.findall('https?://(?:[-\w.]|(?:%[\da-fA-F]{2}))+', text)
    return urls

def get_urls(url):
    urls = []
    try:
        r = requests.get(url, verify=False, timeout=2)
        # TODO
        # Lấy các url nằm trong trang web của url này, lưu lại vào biến urls
        urls = get_url_in_response(r.text)
        return urls
    except:
        print("An exception occurred")
        return []


def insert_url_into_arr(urls, new_urls, limit):
    count = 0;
    for url in new_urls:
        if count == limit:
            break
        if url not in urls:
            urls.append(url)
        count += 1
    return urls

def get_urls_recursive(start_url, limit):
    urls = [start_url]
    tmp_list = []
    for url in urls:
        if url not in tmp_list:
            tmp_list.append(url)
            new_urls = get_urls(url)
            if len(new_urls) > 0:
                insert_url_into_arr(urls,new_urls,limit )
            print(len(urls))
        if len(urls) > 200:
            return urls[0:200]
    return urls
url_list = get_urls_recursive('https://en.wikipedia.org/wiki/Web_mining', 200)
# print(url_list)

42
92
98
98
An exception occurred
98
105
108
114
118
121
125
An exception occurred
125
125
134
156
164
176
182
185
186
186
187
304
['https://en.wikipedia.org/wiki/Web_mining', 'https://en.wikipedia.org', 'https://archive.org', 'http://alexandria.tue.nl', 'http://www.cis.unisa.edu.au', 'https://ieeexplore.ieee.org', 'https://doi.org', 'https://api.semanticscholar.org', 'https://docs.google.com', 'https://ui.adsabs.harvard.edu', 'http://library.ifla.org', 'http://soave.isti.cnr.it', 'http://facweb.cs.depaul.edu', 'https://link.springer.com', 'http://robotics.stanford.edu', 'http://informationr.net', 'https://www.researchgate.net', 'http://dmr.cs.umn.edu', 'http://webmining.spd.louisville.edu', 'http://www.chris-kimble.com', 'http://csdl2.computer.org', 'https://donate.wikimedia.org', 'https://www.wikidata.org', 'https://ar.wikipedia.org', 'https://de.wikipedia.org', 'https://es.wikipedia.org', 'https://eu.wikipedia.org', 'https://fa.wikipedia.org', 'https://fr.wikipedia.org', 'https://ko

## 2. Thu thập dữ liệu
Thu thập dữ liệu từ `url_list`. Lưu trữ dữ liệu thu được vào dictionary data với keys là các từ, values gồm 2 phần tử: 
- `url_idx_list` với $idx \in \left[0,200\right) \cap \mathbb{N}$
- `frequency` 
    
Ví dụ: `data['at']=[url_idx_list,frequency]`:
- `url_idx_list`: danh sách các url mà trong dữ liệu của chúng (html document) chứa từ "at". 
- `frequency`: tần suất xuất hiện (số lần xuất hiện) của từ `at` trong dữ liệu của **tất cả đường dẫn thu được**.   

In [15]:
def text_filter(element):
    if element.parent.name in ['style', 'title', 'script', 'head', '[document]', 'class', 'a', 'li']:
        return False
    elif isinstance(element, Comment):
        '''Opinion mining?'''
        return False
    elif re.match(r"[\s\r\n]+",str(element)): 
        '''space, return, endline'''
        return False
    return True

def wordList(url):
    r =  None
    try:
        r = requests.get(url)
    except:
        print("An exception occurred")
    if r == None:
        return 
    soup = BeautifulSoup(r.content, "html.parser")
    text = soup.findAll(text=True)
    filtered_text = list(filter(text_filter, text)) # list của các chuỗi
    word_list = []
    # print(text)
    translator = str.maketrans('', '', string.punctuation)
    for i in range(len(filtered_text)):
        text_trans = filtered_text[i].translate(translator).strip()
        text_split = text_trans.split(" ")
        for j in range(len(text_split)):
            if len(text_split[j]) != 0:
                word_list.append(text_split[j])
    return ' '.join([str(elem) for elem in word_list])

def read_url(url, url_idx, data):
    word_list = wordList(url)
    print(word_list)
    # if word_list == None:
    #     return
    # # Với mỗi từ w trong word_list:
    # for i in range(len(word_list)):
    #     word = word_list[i]
    #     if word in data.keys():
    #         item = data.get(word)
    #         data[word][1] = data[word][1] + 1
    #         if url_idx not in data[word][0]:
    #             data[word][0].append(url_idx)
    #     else:
    #         item = []
    #         item.append([url_idx]) # data[w]
    #         item.append(1)
    #         data[word_list[i]] = item

In [16]:
data = {}
for url_index, url in enumerate(url_list[:1], 1):
    read_url(url, url_index, data)

{33: None, 34: None, 35: None, 36: None, 37: None, 38: None, 39: None, 40: None, 41: None, 42: None, 43: None, 44: None, 45: None, 46: None, 47: None, 58: None, 59: None, 60: None, 61: None, 62: None, 63: None, 64: None, 91: None, 92: None, 93: None, 94: None, 95: None, 96: None, 123: None, 124: None, 125: None, 126: None}
['Web', 'mining', 'From', 'Wikipedia', 'the', 'free', 'encyclopedia', 'This', 'article', 'includes', 'a', 'related', 'reading', 'or', 'but', 'its', 'sources', 'remain', 'unclear', 'because', 'it', 'lacks', 'October', '2020', 'This', 'article', 'may', 'require', 'The', 'specific', 'problem', 'is', 'The', 'article', 'needs', 'sufficient', 'inline', 'references', 'and', 'a', 'complete', 'rewrite', 'has', 'been', 'proposed', 'October', '2020', 'Web', 'mining', 'It', 'uses', 'automated', 'methods', 'to', 'extract', 'both', 'structured', 'and', 'unstructured', 'data', 'from', 'web', 'pages', 'server', 'logs', 'and', 'link', 'structures', 'There', 'are', 'three', 'main', 's

## 3. Tiền xử lý
Loại bỏ các item trong data mà key là các stopword.

**Ngữ liệu:** 

In [15]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
english_stopwords = stopwords.words('english')
print(english_stopwords)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [18]:
# TODO
# Loại bỏ các key của biến data mà nằm trong danh sách english_stopwords
test_data = data
for stop in english_stopwords:
    if stop in test_data.values():
        test_data.pop(stop, None)
print(test_data)

Bioloji': [[194], 1], 'növlərin': [[194], 1], 'sistematikası': [[194], 1], 'Mediaviki': [[194], 1], 'proqram': [[194], 1], 'təminatının': [[194], 1], 'təkmilləşdirilməsi': [[194], 1], 'Vikiverilənlər': [[194], 1], 'Ortaq': [[194], 2], 'verilənlər': [[194], 1], 'bazası': [[194], 1], 'Vikianbar': [[194], 1], 'anbarı': [[194], 1], 'Vikisəyahət': [[194], 1], 'səyahət': [[194], 1], 'təlimatçası': [[194], 1], 'koordinasiya': [[194], 1], 'mərkəzi': [[194], 1], 'Türkdilli': [[194], 1], 'Vikipediyalar': [[194], 1], 'Mənbə': [[194], 1], 'Naviqasiya': [[194], 1], 'menyusu': [[194], 1], 'Alətlər': [[194], 2], 'sandığı': [[194], 1], 'Adlar': [[194], 1], 'fəzası': [[194], 1], 'Variantlar': [[194], 1], 'Görünüş': [[194], 1], 'çox': [[194], 1], 'Axtar': [[194], 1], 'Bələdçi': [[194], 1], 'Layihələr': [[194], 1], 'Xüsusi': [[194], 1], 'Çap': [[194], 1], 'etixrac': [[194], 1], 'layihələrdə': [[194], 1], 'Başqa': [[194], 1], 'dillərdə': [[194], 1], 'آنا': [[195], 1], 'صفحه': [[195], 1], 'آچیق': [[195], 2

## 4. Lưu trữ và biểu diễn dữ liệu
Sử dụng pickle lưu lại data với tên file raw_data.
### 4.1 Cơ sở dữ liệu giao tác:
Thông thường, các cơ sở dữ liệu giao tác được lưu trong flat files (các tập phẳng) thay vì trong một hệ cơ sở dữ liệu. Các item là các số nguyên không âm, mỗi giao tác tương ứng với một dòng các số nguyên phân tách nhau bằng khoảng trắng.
Ví dụ:

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 

30 31 32 

33 34 35 

36 37 38 39 40 41 42 43 44 45 46 

38 39 47 48 

38 39 48 49 50 51 52 53 54 55 56 57 58 

32 41 59 60 61 62 

3 39 48 

63 64 65 66 67 68 

32 69 

48 70 71 72 

39 73 74 75 76 77 78 79 

36 38 39 41 48 79 80 81 

82 83 84 

41 85 86 87 88 

39 48 89 90 91 92 93 94 95 96 97 98 99 100 101 

36 38 39 48 89 

39 41 102 103 104 105 106 107 108 

38 39 41 109 110 

39 111 112 113 114 115 116 117 118 

119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 

48 134 135 136 

39 48 137 138 139 140 141 142 143 144 145 146 147 148 149 

39 150 151 152 

38 39 56 153 154 155 

In [19]:
# with open('raw_data', 'wb') as f:
#     # TODO
def save_object(obj, file_name):
    with open(file_name, 'wb') as output:  # Overwrites any existing file.
        pickle.dump(obj, output, pickle.HIGHEST_PROTOCOL)
save_object(test_data, "raw_data")

### 4.2 Xuất dataset
Lưu một cơ sở dữ liệu giao tác (transactional database) vào file t_data.txt: 
- Các item tương ứng với url_idx
- Mỗi transaction tương ứng với một từ.

In [1]:
with open('t_data.txt', 'w') as f:
    for word, (url_list, freq) in data.items():
        print(*url_list, file=f)

NameError: name 'data' is not defined

## 5. Truy vấn and, or, not
Ví dụ: 
- Truy vấn `and` câu `web mining`: trả về đường dẫn tới các trang web có cả 2 từ web và từ mining. 
- Truy vấn `or` câu `web mining`: trả về đường dẫn tới các trang web có từ web hoặc từ mining.
- Truy vấn `not` câu `web mining`: trả về đường dẫn tới các trang không có cả từ web và từ mining.

*GỢI Ý: TÁCH CÂU TRUY VẤN THÀNH CÁC TỪ TƯƠNG TỰ PHƯƠNG PHÁP LÀM Ở LAB02.*

In [ ]:
ret=[]
def andRetrieval(ret, sentence):
    '''Parameters
    -----------------------
    ret: url_list
    sentence: query'''
    # TODO
    ### split sentence (separator ' ') into terms
    ### find urls have all terms: urls
    ### if len(ret)==0:  return urls
    ### else update ret with urls: intersection of ret and urls 


def orRetrieval(ret, sentence):
    '''Parameters
    -----------------------
    ret: url_list
    sentence: query'''
    # TODO
    ### split sentence (separator ' ') into terms
    ### find urls have all terms: urls
    ### find urls have at least 1 term: urls
    ### update ret with urls: extend ret with urls

def notRetrieval(ret, sentence):
    '''Parameters
    -----------------------
    ret: url_list
    sentence: query'''
    # TODO
    ### split sentence (separator ' ') into terms
    ### find urls have at least 1 term: urls
    ### update ret with urls: remove urls from ret 

In [ ]:
print(andRetrieval([], 'web mining'))
print(orRetrieval([], 'web mining'))
print(notRetrieval(url_list, 'web mining'))